### PR TITLE
dealing with innerHTML

### PR DESCRIPTION
**Pull request by schomery**

### CODE REVIEW

**Comment by schomery** [comment]
@tophf do we need https://github.com/openstyles/stylus/blob/e48e1ab874e1946cf55a023452de8ded6ff6c2d6/js/localization.js#L26? if not we can throw an error in this block

**Comment by schomery** [comment]
@tophf we don't seem to use HTML codes for the title of messageBox. Can we force textContent then?

**Comment by tophf** [comment]
We don't need this ancient `tE` at all if you replace the three occurrences with e.g. `$('#heading').textContent = t('addStyleTitle');`

**Comment by tophf** [comment]
As for the master branch conflict let's keep your approach in showHelp(), but with `$('.contents', div).textContent = '';` before appendChild.

**Comment by tophf** [comment]
>we don't seem to use HTML codes for the title of messageBox. Can we force textContent then?

Sure.

File: edit/edit.js [hunk 1]
```diff
@@ -260,24 +260,25 @@ function initCodeMirror() {
   };
 
   // initialize global editor controls
-  function optionsHtmlFromArray(options) {
-    return options.map(opt => '<option>' + opt + '</option>').join('');
+  function optionsFromArray(parent, options) {
+    options.map(opt => $element({tag: 'option', textContent: opt}))
+      .forEach(opt => parent.appendChild(opt));
```
Comment (tophf) [inline edit/edit.js]: This function works with a live element (it's in DOM), so multiple appendChild are slower than one, which is why I use DocumentFragment in such cases: it appends the elements as a flat list in one operation.

```js
const fragment = document.createDocumentFragment();
for (const opt of options) {
  fragment.appendChild($element({tag: 'option', textContent: opt}));
}
parent.appendChild(fragment);
```

File: edit/edit.js [hunk 2]
```diff
@@ -314,15 +315,17 @@ function acmeEventListener(event) {
...
+      document.head.appendChild($element({
+        tag: 'link',
+        id: 'cm-theme2',
+        rel: 'stylesheet',
+        href: url
+      }));
+      setTimeout(() => {
+        CodeMirror.setOption(option, value);
+        themeLink.remove();
+        document.getElementById('cm-theme2').id = 'cm-theme';
```
Comment (tophf) [inline edit/edit.js]: Since we touch it, what about `$('#cm-theme2').id = 'cm-theme';`?

File: edit/edit.js [hunk 3]
```diff
@@ -1744,12 +1747,17 @@ function showKeyMapHelp() {
     const col = input.parentNode.cellIndex;
     inputs[1 - col].value = '';
     table.tBodies[0].childNodes.forEach(row => {
-      let cell = row.children[col];
-      cell.innerHTML = cell.textContent.replace(query, '<mark>$&</mark>');
+      const cell = row.children[col];
```
Comment (tophf) [inline edit/edit.js]: Nah, here's the entire block (not tested though):

```js
const cell = row.children[col];
const text = cell.textContent;
const match = text.match(query);
if (match) {
  cell.textContent = text.substr(0, match.index);
  cell.appendChild($element({tag: 'mark', textContent: match[0]}));
  cell.appendChild(document.createTextNode(text.substr(match[0].length + match.index));
} else if (cell.children.length) {
  cell.textContent = text;
}
row.style.display = match ? '' : 'none';
```

File: edit/edit.js [hunk 4]
```diff
@@ -1744,12 +1747,17 @@ function showKeyMapHelp() {
...
+      const html = cell.textContent.replace(query, '<mark>$&</mark>');
+      cell.textContent = '';
+      const div = tHTML(html, 'div');
+      while (div.childNodes.length > 0) {
+        cell.appendChild(div.childNodes[0]);
+      }
       row.style.display = query.test(cell.textContent) ? '' : 'none';
       // clear highlight from the other column
```
Comment (tophf) [inline edit/edit.js]: Why did you remove it?
Should be something like this I think:

```js
const otherCell = row.children[1 - col];
if (otherCell.children.length) {
  otherCell.textContent = otherCell.textContent;
}
```

File: js/localization.js [hunk 5]
```diff
@@ -17,24 +17,23 @@ function t(key, params) {
...
 
 
-function tE(id, key, attr, esc) {
-  if (attr) {
-    document.getElementById(id).setAttribute(attr, t(key));
-  } else if (typeof esc === 'undefined' || esc) {
-    document.getElementById(id).appendChild(document.createTextNode(t(key)));
-  } else {
-    document.getElementById(id).innerHTML = t(key);
+function tHTML(html, tag) {
```
Comment (tophf) [inline js/localization.js]: Judging by AMO [retarded] rules you'll have to rework all HTML string construction properly (completely) into DOM/$element and delete this function.

File: edit/edit.js (hunk 3 above)
Comment (schomery) [inline edit/edit.js]: This is going to highlight a single match not all the matches, right?

File: edit/edit.js (hunk 4 above)
Comment (schomery) [inline edit/edit.js]: I am not following. `otherCell.textContent = otherCell.textContent` does nothing.

File: js/localization.js (hunk 5 above)
Comment (schomery) [inline js/localization.js]: That's right. But still these changes will remove several warnings.

File: edit/edit.js (hunk 4 above)
Comment (tophf) [inline edit/edit.js]: It removes child element (`<mark>`)

File: edit/edit.js (hunk 3 above)
Comment (tophf) [inline edit/edit.js]: Oops... indeed. Well, it can be implemented via string#split or regexp#exec. Someday you'll have to do it anyway (due to AMO) so it's up to you.

File: edit/edit.js (hunk 2 above)
Comment (schomery) [inline edit/edit.js]: done

File: edit/edit.js (hunk 1 above)
Comment (schomery) [inline edit/edit.js]: done

File: edit/edit.js (hunk 3 above)
Comment (schomery) [inline edit/edit.js]: done

File: edit/edit.js (hunk 4 above)
Comment (schomery) [inline edit/edit.js]: True!

**Comment by narcolepticinsomniac** [comment]
I'm using this version for testing. Importing Mozilla format results in an empty "issues" popup. Suppose it could be from either PR.

**Comment by tophf** [comment]
> an empty "issues" popup

Mea culpa!
Fixed in 02fa1fb.

**Comment by narcolepticinsomniac** [comment]
Our html is poking out for the whole world to see. On the manage page:

manage

File: manage.html [hunk 6]
```diff
@@ -212,7 +212,7 @@ <h2 id="backup-title" i18n-text="backupButtons"></h2>
       <button id="unfile-all-styles" i18n-text="retrieveBckp"></button>
     </p>
   </div>
-  <p id="manage-text" i18n-html="manageText"></p>
+  <p id="manage-text" i18n-text="manageText"></p>
```
Comment (tophf) [inline manage.html]: I was wrong: this one shouldn't be changed

**Comment by tophf** [comment]
@schomery, I've added a fixup commit for the manageText. When working in a PR branch, I always rebase and squash such trivial fixups to an offending earlier commit in the same branch, then `git push -f origin branchname`. WDYT?

**Comment by narcolepticinsomniac** [comment]
Beautifier is completely broken. I think this is the related error:

beautify

You can also see in the pic that that stupid `tabId` error is back. I know it's inconsequential, but it had been a long time since it popped up, which was nice.

**Comment by tophf** [comment]
The beautifier is fixed now. As for tabId, I've added another check. Anyway, the error may be safely ignored, and normal users won't see it on chrome://extensions page anyway.

**Comment by narcolepticinsomniac** [comment]
Cool.

Something else a little weird, although I'm not sure that it's new, is CodeMirror's failure to colorize large pastes properly. It works till a certain amount of characters, then fails. I have seen this in the past, but it was always absurd amounts of text. This time, I was copy/pasting a sites optional dark theme from its .css file. It's a fair amount of text, and it's mostly minified, so it's a wall. If you break it in half though, with a blank line and paste it like that, colors work right. 

Examples:

Colors fail

[paste example.txt](https://github.com/openstyles/stylus/files/1167117/paste.example.txt)

Colors work

[paste example split.txt](https://github.com/openstyles/stylus/files/1167118/paste.example.split.txt)

**Comment by tophf** [comment]
I'm not surprised CodeMirror stops colorizing lines of o15k chars. It could be smarter of course but this is too much of an extreme and ridiculous edge case for an online code editor.

**Comment by narcolepticinsomniac** [comment]
Fair enough. I just find it odd that you can paste the same exact amount of text at once, and as long as there's a newline, colors work as expected. Not saying it's a huge deal, just weird.

**Comment by tophf** [comment]
The editors are processing/colorizing code line by line so the important thing is the length of a line, not the overall amount of text.

**Comment by narcolepticinsomniac** [comment]
Got it. Don't paste bricks. wontfix.

**Comment by tophf** [comment]
Well, it's a CodeMirror's issue, so dunno about their wontfixing it. Maybe they will... You can check their repo for an existing bug report or submit a new one using their [online demo](https://codemirror.net/demo/complete.html) and some minified javascript from e.g. google search page etc.

**Comment by schomery** [comment]
> When working in a PR branch, I always rebase and squash such trivial fixups to an offending earlier commit in the same branch

@tophf Sure. The list will be cleaner.

**Comment by tophf** [comment]
Huh, I did the rebase on master but now each commit has me as an additional author... I guess we shouldn't rebase on other's branches. Aside from this weirdness everything is fine: code diff is empty for the rebased branch against the original one.

**Comment by schomery** [comment]
Looks good to me. If this gets merged to the master branch we can resubmit the extension to AMO and see their reaction to the changes. @narcolepticinsomniac do you have any other comments?

**Comment by narcolepticinsomniac** [comment]
Tophf squashed the bugs I've come across. Nothing major that should prevent you from submitting, although it'd be good if you could check in a couple times this week to make sure nothing else comes up.

**Comment by Mottie** [comment]
It all looks good to me too... the only odd thing I ran into was the emacs keyboard shortcuts, but I don't think that is related to the innerHTML branch.

2017-07-22 07_51_08-edit style stackoverflow dark

**Comment by tophf** [comment]
@Mottie, isn't it the way it always worked? IIRC this is intentional, because that's how the keymap is defined in emacs.

**Comment by Mottie** [comment]
LOL ok... I haven't used emacs before, so looking at the information provided doesn't seem very useful.

**Comment by tophf** [comment]
Emacs is a weird thing all around.
